### DISCUSSION TITLE
SDK v0.4

### DISCUSSION THREAD

**Discussion by karolkielecki** [saleor/saleor#8843] [Show and tell]
We've recently released v0.4 of our SDK!

Authentication is now fully supported. We are successfully using it in production for some of our projects. Feel free to check it out and let us know your thoughts!

https://www.npmjs.com/package/@saleor/sdk

---
[Twitter](https://twitter.com/getsaleor/status/1472870478775791620)

**Comment by haneef-photon** [comment]
Can you help me with the steps to integrate external oAuth providers using the SDK? I intend to use Firebase authentication as my authentication system and Saleor as a resource server. As per the documentation, "_To use this flow you need to provide values for JSON Web Key Set URL and User info URL on the Saleor Dashboard side."_  I understand that I would need the OIDC plugin (which currently is only a cloud offering) but is there a way to set these settings in Saleor core or dashboard so that I can use the token provided by Firebase and use it to fetch user data such as orders etc? If not, what is the best way to handle this?

Since I am not familiar with python/Django, I was thinking of a crude way to implement this: 
1. In the front-end, login using Firebase auth, create a random user password and save it in firebase user's account.
2. Once logged in successfully, create user account in Saleor with the newly created password and log them in.

**Reply by haneef-photon** [reply]
> Can you help me with the steps to integrate external oAuth providers using the SDK? I intend to use Firebase authentication as my authentication system and Saleor as a resource server. As per the documentation, "_To use this flow you need to provide values for JSON Web Key Set URL and User info URL on the Saleor Dashboard side."_ I understand that I would need the OIDC plugin (which currently is only a cloud offering) but is there a way to set these settings in Saleor core or dashboard so that I can use the token provided by Firebase and use it to fetch user data such as orders etc? If not, what is the best way to handle this?
> 
> Since I am not familiar with python/Django, I was thinking of a crude way to implement this:
> 
>     1. In the front-end, login using Firebase auth, create a random user password and save it in firebase user's account.
> 
>     2. Once logged in successfully, create user account in Saleor with the newly created password and log them in.

Here is an update from my further research, @karolkielecki can you let me know if this is the correct flow that I can follow to get things working?:

I took a look at the latest SDK, from what I was able to understand is, for the OIDC resource server flow, these are the steps that are required:

1. At the client (storefront), need to use the Auth,` externalVerify()` method to pass the pluginID, token to the server.
2. Create a new plugin and implement the `user_authenticate() `method which accepts the token, parse and verify the token and return the `User` object associated with the token.

Hopefully this is all that is required from implementation part for the resource server flow. Feel free to correct me if I missed something and did not understand the requirements for this implementation correctly.

**Reply by karolkielecki** [reply]
Hi @haneef-photon, I'll pass it to my colleague. We'll get back to you soon.

**Reply by korycins** [reply]
Hey, @haneef-photon,
Saleor has an interface for building an external auth interface (for example, this is used by OIDC plugin). IMHO you need to build own plugin, that will handle the external authorization of the user.

**Reply by haneef-photon** [reply]
Based on my understanding on what is needed to be implemented from be Base plugin class, I wanted to confirm if `user_authenticate()` is the only method that I need to implement for the resource server flow? What will be the inputs to this method and what should I return in response?

From the front-end, to verify the JWT token received from external provider such as firebase, can I use `exernalVerify()` method from the SDK that will log the user in?

**Reply by korycins** [reply]
You need to go through the OIDC docs as an example and check what mutations are used there. All of them are connected with plugins' hooks. I assume that all of this methods should be implemented by you.

**Reply by haneef-photon** [reply]
Thanks, let me try
